### PR TITLE
Abstract Conditional for Hybrid

### PR DESCRIPTION
**Pull request by ProfFan**
This PR implements the idea of making `Conditional` base on a concrete `AbstractConditional` so we can using virtual dispatch.

All tests pass except one - hybrid now actually does hybrid, so the full elimination fails, as we didn't implement discrete elimination in `EliminateHybrid`:

```
testHybridFactorGraph
gtsam/hybrid/tests/testHybridFactorGraph.cpp:377: Failure: "Exception: DecisionTree::convertFrom: Invalid NodePtr" 
There were 1 failures

Process finished with exit code 1
```

### CODE REVIEW

File: gtsam/hybrid/tests/testHybridFactorGraph.cpp [hunk 1]
```diff
@@ -365,13 +365,50 @@ TEST(HybridFactorGraph, Elimination) {
...
+  auto linearizedFactorGraph = self.linearizedFactorGraph;
+
+  // Create ordering.
+  Ordering ordering;
+  for (size_t k = 1; k <= self.K; k++) ordering += X(k);
+  for (size_t k = 1; k < self.K; k++) ordering += M(k);
+
+  // Eliminate partially.
+  HybridBayesNet::shared_ptr hybridBayesNet =
+      linearizedFactorGraph.eliminateSequential(ordering);
```
Comment (ProfFan) [inline gtsam/hybrid/tests/testHybridFactorGraph.cpp:389]: Fails here

File: gtsam/hybrid/HybridBayesNet.cpp [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+#include <gtsam/hybrid/HybridBayesNet.h>
+
+namespace gtsam {
+
+GaussianMixture::shared_ptr HybridBayesNet::atGaussian(size_t i) {
+  return boost::dynamic_pointer_cast<GaussianMixture>(factors_.at(i));
+}
+
+}
```
Comment (dellaert) [inline gtsam/hybrid/HybridBayesNet.cpp:13]: return

File: gtsam/hybrid/HybridBayesNet.cpp [hunk 3]
```diff
@@ -0,0 +1,13 @@
+//
+// Created by Fan Jiang on 1/24/22.
+//
+
+#include <gtsam/hybrid/HybridBayesNet.h>
+
+namespace gtsam {
+
+GaussianMixture::shared_ptr HybridBayesNet::atGaussian(size_t i) {
```
Comment (dellaert) [inline gtsam/hybrid/HybridBayesNet.cpp:9]: should this not be an abstract conditional ?

File: gtsam/hybrid/tests/testHybridBayesNet.cpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+using symbol_shorthand::X;
+
+static const DiscreteKey Asia(0, 2);
+
+/* ****************************************************************************/
+// Test creation
+TEST(HybridBayesNet, Creation) {
+  HybridBayesNet bayesNet;
+
+  bayesNet.add(Asia, "99/1");
```
Comment (dellaert) [inline gtsam/hybrid/tests/testHybridBayesNet.cpp:39]: Well, add a test and implement :-)

**Review by dellaert — APPROVED** [review]

File: gtsam/hybrid/HybridBayesNet.cpp (hunk 3 above)
Comment (ProfFan) [reply gtsam/hybrid/HybridBayesNet.cpp:9]: This is `atGaussian`, the `at()` does the abstract one

File: gtsam/hybrid/tests/testHybridBayesNet.cpp (hunk 4 above)
Comment (ProfFan) [reply gtsam/hybrid/tests/testHybridBayesNet.cpp:39]: done

**Pull request merged**